src/darwin/Framework/CHIP/MTRLogging.mm

```diff
@@ -0,0 +1,184 @@
+        os_log_fault(OS_LOG_DEFAULT, "%@", message);
+        break;
+    case MTRLogLevelError:
+        os_log_error(OS_LOG_DEFAULT, "%@", message);
+        break;
+    case MTRLogLevelInfo:
+        os_log_info(OS_LOG_DEFAULT, "%@", message);
+        break;
+    case MTRLogLevelDebug:
+        os_log_debug(OS_LOG_DEFAULT, "%@", message);
+        break;
+    default:
+        break;
+    }
+}
+
+@end
+
+// MARK: - MTRLogging
+
+@implementation MTRLoggingConfiguration
+
+static id<MTRLogger> singleton_logger;
+
+/**
+ * A log redirection function that will be used to capture logs from the Matter C++ SDK.
+ */
+static ENFORCE_FORMAT(3, 0) void cpp_log_redirect_callback(const char * module, uint8_t category, const char * msg, va_list args)
+{
+    // Pre-format the message. Additional information will be prepended and appended before
+    // logging.
+    char formatted_msg[CHIP_CONFIG_LOG_MESSAGE_MAX_SIZE];
+    int32_t formatted_msg_length = vsnprintf(formatted_msg, sizeof(formatted_msg), msg, args);
+    if (formatted_msg_length < 0) {
+        // In the event that we could not format the message, simply copy over the
+        // unformatted message. It won't contain all the information that we need, but
+        // it's better than dropping a log on the floor.
+        strncpy(formatted_msg, msg, sizeof(formatted_msg));
+        return;
+    }
+
+    // Explicitly NULL-terminate the formatted string for safety.
+    formatted_msg[sizeof(formatted_msg) - 1] = '\0';
+
+    // Collect the process and thread information to match the original CHIPLogging implementation
+    // in src/platform/Darwin/Logging.cpp
+    uint64_t ktid;
+    pthread_threadid_np(NULL, &ktid);
+    uint64_t pid = getpid();
+
+    // Delegate the log to the MTRLogging system
+    switch (category) {
+    case chip::Logging::LogCategory::kLogCategory_Error:
+        [MTRLogging logAtLevel:MTRLogLevelError
+                        format:@"%@ [%lld:%lld] CHIP: [%s] %s", @"🔴", (long long) pid, (long long) ktid, module, formatted_msg];
+        break;
+
+    case chip::Logging::LogCategory::kLogCategory_Progress:
+        [MTRLogging logAtLevel:MTRLogLevelInfo
+                        format:@"%@ [%lld:%lld] CHIP: [%s] %s", @"🔵", (long long) pid, (long long) ktid, module, formatted_msg];
+        break;
+
+    case chip::Logging::LogCategory::kLogCategory_Detail:
+        [MTRLogging logAtLevel:MTRLogLevelDebug
+                        format:@"%@ [%lld:%lld] CHIP: [%s] %s", @"🟢", (long long) pid, (long long) ktid, module, formatted_msg];
+        break;
+    default:
+        [MTRLogging logAtLevel:MTRLogLevelInfo
+                        format:@"[%lld:%lld] CHIP: [%s] %s", (long long) pid, (long long) ktid, module, formatted_msg];
+        break;
+    }
+}
+
++ (void)setLogger:(id<MTRLogger>)logger
+{
+    singleton_logger = logger;
+    chip::Logging::SetLogRedirectCallback(cpp_log_redirect_callback);
+}
+
++ (id<MTRLogger>)getLogger
+{
+    return singleton_logger;
```

Comment:
This can return nil (if no logger was set), but the API definition promises to never return nil here.

Reply:
Set a default in case it is null